module/planning/PlanKick/src/PlanKick.cpp

```diff
@@ -0,0 +1,103 @@
+            cfg.ball_distance_threshold = config["ball_distance_threshold"].as<float>();
+            cfg.ball_angle_threshold    = config["ball_angle_threshold"].as<float>();
+            cfg.target_angle_threshold  = config["target_angle_threshold"].as<Expression>();
+            cfg.kick_leg                = config["kick_leg"].as<std::string>();
+        });
+
+        on<Provide<KickTo>, Uses<Kick>, Trigger<FilteredBall>, With<Stability>>().then([this](
+                                                                                           const KickTo& kick_to,
+                                                                                           const Uses<Kick>& kick,
+                                                                                           const FilteredBall& ball,
+                                                                                           const Stability& stability) {
+            // If the kick is running, don't interrupt or the robot may fall
+            if (kick.run_state == GroupInfo::RunState::RUNNING && !kick.done) {
+                emit<Task>(std::make_unique<Idle>());
+                return;
+            }
+
+            // CHECK IF BALL IS BALL MEASUREMENT IS RECENT ENOUGH
+            // If the ball measurement is old, then don't do anything
+            auto time_difference =
+                std::chrono::duration_cast<std::chrono::milliseconds>(NUClear::clock::now() - ball.time_of_measurement);
+            if (time_difference.count() >= cfg.ball_timeout_threshold) {
+                return;
+            }
+
+            // CHECK IF CLOSE TO BALL
+            // Get the angle and distance to the ball
+            float ball_angle    = std::abs(std::atan2(ball.rBTt.y(), ball.rBTt.x()));
+            float ball_distance = ball.rBTt.head(2).norm();
+
+            // Need to be near the ball to consider kicking it
+            if (ball_distance > cfg.ball_distance_threshold || ball_angle > cfg.ball_angle_threshold) {
+                return;
+            }
+
+            // CHECK IF FACING POINT TO KICK TO
+            float align_angle = std::abs(std::atan2(kick_to.rPTt.y(), kick_to.rPTt.x()));
+
+            // Don't kick if we should align but we're not aligned to the target
+            if (align_angle > cfg.target_angle_threshold) {
+                return;
+            }
+
+            // If the kick conditions are not met, the function will have returned with no Tasks, ending the kick
+            // Otherwise, the kick conditions are met and we need to check if we are already kicking
+            // If we are already queued to kick, then only emit Idle to keep the previous Kick Task running
+            if (kick.run_state == GroupInfo::RunState::QUEUED) {
+                emit<Task>(std::make_unique<Idle>());
+                return;
+            }
+
+            // If the robot is not standing, make it stand before kicking
+            if (stability != Stability::STANDING) {
```

Comment:
One day... pushing will get rid of this, I'm glad there is this workaround for now though
Manual pushing!